File: src/main/java/com/jaffa/rpc/lib/serialization/KryoPoolSerializer.java
package com.jaffa.rpc.lib.serialization;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import com.esotericsoftware.kryo.util.Pool;
import lombok.extern.slf4j.Slf4j;

import java.io.ByteArrayOutputStream;


@Slf4j
@SuppressWarnings("squid:S1168")
public final class KryoPoolSerializer implements ObjectSerializer {
    private final Pool<Kryo> pool;

    public KryoPoolSerializer() {
        pool = new Pool<Kryo>(true, true, 100) {
            protected Kryo create() {
                Kryo kryo = new Kryo();
                kryo.register(void.class);
                kryo.setRegistrationRequired(false);
                kryo.setReferences(true);
                return kryo;
            }
        };
    }

    @Override
    public byte[] serialize(Object obj) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        Output output = new Output(baos);
        Kryo kryo = pool.obtain();
        kryo.writeObject(output, obj);
        output.flush();
        output.close();
        pool.free(kryo);
        return baos.toByteArray();
    }

    @Override
    public byte[] serializeWithClass(Object obj) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        Output output = new Output(baos);
        Kryo kryo = pool.obtain();
        kryo.writeClassAndObject(output, obj);
        output.flush();
        output.close();
        pool.free(kryo);
        return baos.toByteArray();
    }

    @Override
    public Object deserializeWithClass(byte[] serialized) {
        Object obj;
        Kryo kryo = pool.obtain();
        Input input = new Input(serialized);
        obj = kryo.readClassAndObject(input);
        pool.free(kryo);
        return obj;
    }

    @Override
    public <T> T deserialize(byte[] serialized, Class<T> clazz) {
        T obj;
        Kryo kryo = pool.obtain();
        Input input = new Input(serialized);
        obj = kryo.readObject(input, clazz);
        pool.free(kryo);
        return obj;
    }
}

File: src/main/kotlin/com/jaffa/rpc/lib/serialization/KryoPoolSerializer.kt
package com.jaffa.rpc.lib.serialization

import com.esotericsoftware.kryo.Kryo
import com.esotericsoftware.kryo.io.Input
import com.esotericsoftware.kryo.io.Output
import com.esotericsoftware.kryo.util.Pool
import java.io.ByteArrayOutputStream

class KryoPoolSerializer : ObjectSerializer {
    private val pool: Pool<Kryo>
    override fun serialize(obj: Any?): ByteArray? {
        val baos = ByteArrayOutputStream()
        val output = Output(baos)
        val kryo = pool.obtain()
        kryo.writeObject(output, obj)
        output.flush()
        output.close()
        pool.free(kryo)
        return baos.toByteArray()
    }

    override fun serializeWithClass(obj: Any?): ByteArray? {
        val baos = ByteArrayOutputStream()
        val output = Output(baos)
        val kryo = pool.obtain()
        kryo.writeClassAndObject(output, obj)
        output.flush()
        output.close()
        pool.free(kryo)
        return baos.toByteArray()
    }

    override fun deserializeWithClass(serialized: ByteArray?): Any? {
        val obj: Any?
        val kryo = pool.obtain()
        val input = Input(serialized)
        obj = kryo.readClassAndObject(input)
        pool.free(kryo)
        return obj
    }

    override fun <T> deserialize(serialized: ByteArray?, clazz: Class<T>?): T? {
        val obj: T?
        val kryo = pool.obtain()
        val input = Input(serialized)
        obj = kryo.readObject(input, clazz)
        pool.free(kryo)
        return obj
    }

    init {
        pool = object : Pool<Kryo>(true, true, 100) {
            override fun create(): Kryo {
                return Kryo().also { it.register(Void.TYPE) }.also { it.isRegistrationRequired = true }
                    .also { it.references = true }
            }
        }
    }
}